base/weakkeydict.jl

```diff
@@ -69,56 +64,144 @@ function WeakKeyDict(kv)
     end
 end
 
+function _cleanup_locked(h::WeakKeyDict)
+    if h.dirty
+        h.dirty = false
+        idx = skip_deleted_floor!(h.ht)
+        while idx != 0
+            if h.ht.keys[idx].value === nothing
+                _delete!(h.ht, idx)
+            end
+            idx = skip_deleted(h.ht, idx + 1)
+        end
+    end
+    return h
+end
+
 sizehint!(d::WeakKeyDict, newsz) = sizehint!(d.ht, newsz)
 empty(d::WeakKeyDict, ::Type{K}, ::Type{V}) where {K, V} = WeakKeyDict{K, V}()
 
+IteratorSize(::Type{<:WeakKeyDict}) = SizeUnknown()
+
 islocked(wkh::WeakKeyDict) = islocked(wkh.lock)
 lock(f, wkh::WeakKeyDict) = lock(f, wkh.lock)
 trylock(f, wkh::WeakKeyDict) = trylock(f, wkh.lock)
 
 function setindex!(wkh::WeakKeyDict{K}, v, key) where K
     !isa(key, K) && throw(ArgumentError("$(limitrepr(key)) is not a valid key for type $K"))
-    finalizer(wkh.finalizer, key)
+    # 'nothing' is not valid both because 'finalizer' will reject it,
+    # and because we therefore use it as a sentinel value
+    key === nothing && throw(ArgumentError("`nothing` is not a valid key for type WeakKeyDict"))
```

Comment:
```suggestion
    key === nothing && throw(ArgumentError("`nothing` is not a valid WeakKeyDict key"))
```
Only because the other similar messages refer to the element type instead of the collection type.